keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,343 @@
+
+#### Why gRPC API?
+* gRPC API is strongly typed and can be versioned.
+* Http ReadOnly port has security issues and is disabled in many clusters. The Authenticated port requires authorization by kube-API (the credential are cached in kubelet by default 2minutes) this add the ControlPlane dependency to the API.
+* Using a unix socket we enforce the client to be a workload running on the node.
+
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `podinfo` in Kubelet, listening on a unix socket at `/var/lib/kubelet/pod-info/kubelet.sock`. The endpoint will be versioned.
+The gRPC Service will expose 3 methods serving local pods statuses data:
+
+```protobuf
+service PodInfo {
+    // ListPodConditions returns a stream of List of PodConditions
+    rpc ListPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // GetPodConditions returns a PodInfo for given pod's UID
+    rpc GetPodConditions(GetPodConditionsRequest) returns (PodInfo) {}
+}
+
+// Conditions aligns with v1.PodCondition
+message Conditions {
+      string Type = 1;
+      string Status = 2;
+      Timestamp LastProbeTime = 3;
+      Timestamp LastTransitionTime = 4;
+      string Reason = 5;
+      string Message = 6;
+}
+
+// PodInfo returns a Pod details and list of status Conditions
+message PodInfo {
+    string podUID = 1;
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated Conditions conditions = 5;
+}
+
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
+	repeated PodInfo pods = 1;
+}
+
+// GetPodConditionsRequest contains pods UID
+message GetPodConditionsRequest {
+    string podUID = 1;
+}
+```
+
+##### Unit tests
+
+
+##### e2e tests
+
+  - Feature gate enable/disable tests.
+  - Test API work when kubelet restarts.
+  - Test API work when Control Plane is unreachable.
+
+### Graduation Criteria
+
+#### Alpha
+
+- [ ] Feature implemented behind a feature flag.
+- [ ] e2e tests completed and enabled.
+
+#### Beta
+- [ ] Fix kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277)
+
+#### GA
+
+- [ ] No issue related to usage of API.
```

Comment:
you may likely want to have a bit more precise criterias here (e.g. no major bugs in X months/cycles, or anything you see fit). May be fixed later.